OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -236,15 +238,39 @@ void ITick.Tick(Actor self)
 
 		protected virtual void Tick(Actor self)
 		{
-			while (queue.Count > 0 && BuildableItems().All(b => b.Name != queue[0].Item))
+			foreach (var item in queue.ToArray())
 			{
-				// Refund what's been paid so far
-				playerResources.GiveCash(queue[0].TotalCost - queue[0].RemainingCost);
-				FinishProduction();
+				if (BuildableItems().All(b => b.Name != item.Item))
+				{
+					// Refund what's been paid so far
+					playerResources.GiveCash(item.TotalCost - item.RemainingCost);
+					FinishProduction(item.Item);
+				}
 			}
 
 			if (queue.Count > 0)
-				queue[0].Tick(playerResources);
+			{
+				if (!Info.ParallelBuild)
+				{
+					queue[0].Tick(playerResources, ProductionItem.TimeFactor);
+					return;
+				}
+
+				var progressingQueue = new List<ProductionItem>();
+
+				foreach (var q in queue)
+					if (progressingQueue.Find(i => i.Item == q.Item) == null && !q.Paused)
```

Comment:
Please have !q.Paused be the first item in the if statement as it is cheaper to compute